### ISSUE TITLE
Add a monotonic clock to the framework

### ISSUE THREAD

**Issue by GSPP** [dotnet/runtime#15207]
**Problem:** Many BCL components use `DateTime.UtcNow` to compute timeouts and time deltas. When you run a "Find Usages" on `DateTime.UtcNow` in .NET Framework assemblies a lot of usage sites come to light. For example, in ADO.NET, transactions, caches, WebRequest, remoting, WCF, SignalR, security.

**All of these usages are bugs** because the current system time can change significantly forwards and backwards at any time. This means that in case the system clock is changed many .NET components will spuriously fail.

Common symptoms would be timeouts firing to early or too late (possibly never). Perf counters also might show too high or negative values.

Sample scenario:
1. The system clock spuriously jumps forward due to misconfiguration by years.
2. A cache item is added.
3. The clock is fixed.

Then, the cache item will essentially stay around forever if eviction is based on `DateTime.UtcNow`. I believe this bug exists in the BCL.

User code also has the same problems.

**Solution:** Add some kind of monotonic clock to the BCL. The main property of such a clock would be that its time advances linearly. It should never jump when the system clock changes. It should not exhibit split second jumps. It should behave reasonably in case the system sleeps or hibernates or in case the VM is paused.

This monotonic clock class should be fast and convenient to use so that it is a no-brainer to switch all `DateTime.UtcNow` usages over to the new model.

Maybe we can add `Environment.TickCount64` as well. `Environment.TickCount64` alone would not be good enough because it is awkward to use.

I'm not sure if `Stopwatch` would cover all these requirements. It is a reference type which might be too heavy for hot code path in the BCL.

Maybe we can add a new value type that is made exactly for this purpose. A sketch:

```
struct MonotonicTime {
 long Ticks;
 MonotonicTime operator + (MonotonicTime, TimeSpan);
 TimeSpan operator - (MonotonicTime, MonotonicTime);
 bool IsNegative;
 static MonotonicTime Current;
 ...
}
```

To summarize, I request:
1. The framework timeouts should be fixed. (The TPL bugs apparently are already fixed. I reported these a while ago.)
2. Add `Environment.TickCount64`.
3. Provide a convenient, high-performance monotonic clock.

**Comment by GrabYourPitchforks** [comment]
I agree with an earlier post that said `DateTime.UtcNow` is appropriate for the overwhelming majority of callers. Most applications are resilient to the system clock changing by a few seconds here and there. If a system administrator has applications that are not resilient to this, then the sysadmin should disable the system time syncing service.

I don't think we need to worry about the case where a misconfiguration sends the system clock forward or backward by years. That would represent a system-wide misconfiguration and would hopefully be caught very quickly. And the recovery procedure is simple enough: fix the clock and reboot the OS.

IMO this API is niche enough where it doesn't need to be in-box. Anybody who consumes it probably has their own opinionated beliefs regarding how it should behave.

**Comment by Tragetaschen** [comment]
Not necessarily arguing against the niche, but reducing this to the system administrator and a "just reboot" attitude is a bit short sighted.

I'm coming over from https://github.com/dotnet/aspnetcore/issues/13628 where an embedded hardware has an unreliable system clock due to design limitations. Changing it manually or automatically on the order of O(months) is not uncommon and results in unexpected error conditions due to the usage of `DateTime.UtcNow` within ASP.NET Core. I agree with the OPs sentiment: `All of these usages are bugs`.
While `.UtcNow` is orders of magnitude better compared to `.Now`, I have seen countless real-life issues with programs using the (equivalent of the) latter. Our test engineer has pretty much given up on running any kind of measurements over weekends with a DST change…

If the monotonic clock API is too niche, I would argue that having a dedicated API for timeouts (based on a monotonic time source) is a valuable addition to the framework. Instead of pointing users of `.Now` to the still-not-fully-correct `.UtcNow`, this would provide an actual solution to implement timeout correctly under all circumstances.

**Comment by GSPP** [comment]
Here are two alternative designs:

<hr>

```
class Environment
{
    public static TimeSpan TimeSinceProcessStart { get; }
}
```
`TimeSinceProcessStart` would internally be based on stopwatch time stamps. It would be a very simple mechanism. This time value could be useful for other diagnostic purposes. If we want to initialize this lazily then it needs to be renamed `ContinuousTime` because it no longer starts when the process starts.

<hr>

```
class Timeout
{
    static Timeout Create(TimeSpan delay);

    public TimeSpan TimeRemaining { get; }
    public bool IsElapsed { get; }
    public DateTime ElapseDateTimeUtc { get; }

    public IDisposable WhenElapsed(Action action);
    public Task WhenElapsedAsync();
    public void Propagate(CancellationTokenSource cts);
}
```

<hr>

I'm personally not convinced that these should be added to the framework. I'm logging the ideas here to collect them.

**Comment by Clockwork-Muse** [comment]
> `public DateTime ElapseDateTimeUtc { get; }`

Strictly speaking, a date/time instant can't be elapsed, because it's not a duration.  You can have a "target" instant (the estimated instant the event will fire), and you can have the instant it _actually_ fired; these are, however, not guaranteed to be the same value.  Which of them did you mean?

Things get even crazier in the (admittedly rare) case of the computer clock being manually adjusted for whatever reason (kernels tend to be better behaved about the adjustments they perform on the clocks).  For this reason, low-level timeout functionality should almost never make reference to the "current" ("real world") time, but should instead be based and report solely on elapsed process/system time.

**Comment by GSPP** [comment]
I meant the elapse target. The elapse target could change measured in the currently configured PC clock time. It's meant as a supplementary property. Not sure about the use case, I just added it as an idea. This proposal was done rather quickly...

The purpose of a timeout usually is to elapse after a certain absolute wall-clock duration has elapsed. "We have 5 seconds to make this HTTP call or else the user is going to be delayed for too long.". This type should encapsulate such semantics.

**Comment by Clockwork-Muse** [comment]
> The purpose of a timeout usually is to elapse after a certain absolute wall-clock duration has elapsed. "We have 5 seconds to make this HTTP call or else the user is going to be delayed for too long.". This type should encapsulate such semantics.

Right, then I recommend removing it.  Referencing wall-clock time is more the domain of a scheduling process, which is a way higher level concern.

**Comment by GrabYourPitchforks** [comment]
How often do your production systems adjust their system clocks so significantly that apps require special handling of this condition?

**Comment by Clockwork-Muse** [comment]
> How often do your production systems adjust their system clocks so significantly that apps require special handling of this condition?
>
> Edit: this question is in relation to exposing new API. Let's assume for now that we can make Timer and other types more resilient to system clock changes without exposing any new API surface.

Strictly speaking, every piece of **hardware** will have the clock significantly updated at least once - usually during install, when the clock is first set.  In most cases this is going to be the only time such an update happens (barring failure of the clock battery when powered off).  Updates from NTP and similar services are usually better behaved, specifically to avoid some of these problems.  Changing the _time zone_ of the system does not update the clock (or should not, in a modern OS).

QPC and similar system (or process) time clocks should be immune to time changes, because they measure the relative ticks since a start, and aren't based around an external absolute instant (ie, UTC).  Outside of some oddball situations for fallbacks, then, `Stopwatch` should already be immune, and exposes an `Elapsed` property.

The question, then, is more whether we want a system/process `Elapsed` property, which can be based off the same (and is more convenient than having to deal with `new`-ing something up), like @GSPP has proposed:

```csharp
class Environment
{
    public static TimeSpan TimeSinceProcessStart { get; }
}
```

**Comment by ayende** [comment]
Just to note on the frequency. That is actually happening quite often.
In the cloud, you can "pause" instances and there is physical host migration as well. That can cause the system to freeze for a while.

It used to be that we would see clock drift that exceeded 15 minutes over a period of a few months (which caused sync issues with APIs). 

If you go to https://time.gov/, you can see your system's offset from the official time. My laptop is off by +0.040 s, for example.

**Comment by GSPP** [comment]
Adjusting the clock is a rare event. But when it happens we don't want random malfunctions. OK, I don't want to overstate the significance of this problem. But it's the kind of totally unexpected bug that takes down, say, Azure in the middle of the night.

**Comment by Clockwork-Muse** [comment]
>Just to note on the frequency. That is actually happening quite often.
>In the cloud, you can "pause" instances and there is physical host migration as well. That can cause the system to freeze for a while.
>
>It used to be that we would see clock drift that exceeded 15 minutes over a period of a few months (which caused sync issues with APIs).
>
>If you go to https://time.gov/, you can see your system's offset from the official time. My laptop is off by +0.040 s, for example.

These are all instances of absolute time differences.  Although normally I would have expected things like host migration to be "invisible", because I would have expected the host clock to be correct due to management requirements.  Clock drift over months tells me there's a configuration issue, because NTP updates way more often than that by default.  Also, that's a **spectacularly bad** accuracy - your common wristwatch is something like +-30 seconds a month (and at 15 minutes off, most OSs will start squawking at you for SSL connections).

> It would be nice to have a correct-by-default solution that is clean and convenient.
>
>(I updated the class Timeout slightly.)

@GSPP - in what way is the existing `Timer` class(es) insufficient?
I have misgivings about this method, regardless:
```csharp
static Timeout CreateFromDeadline(DateTime elapseTargetDateTimeUtc);
```
... because it's going to be based on a lie (that the timeout is actually based on absolute time).

**Comment by tannergooding** [comment]
Closing this as there isn't currently anything actionable.

.NET provides multiple times and time sources. Dealing with time, is and always will be complicated and there is no "foolproof" solution.

If you want a monotonic clock then `Stopwatch` is your best bet. You can easily convert it to a `TimeSpan` for use in other scenarios and it isn't subject to arbitrary changes in time due to accidental user misconfiguration/etc.
